spec/features/admin/bulk_product_update_spec.rb

```diff
@@ -744,4 +744,79 @@
       expect(v.on_hand).to eq 18
     end
   end
+
+  describe "Updating product image with new upload interface" do
+    let!(:product) { create(:simple_product, name: "Carrots") }
+
+    it "displays product images and image upload modal" do
+      quick_login_as_admin
+      visit '/admin/products/bulk_edit'
+
+      within "table#listing_products tr#p_#{product.id}" do
+        # Displays product images
+        expect(page).to have_selector "td.image"
+
+        # Shows default image when no image set
+        expect(page).to have_css "img[src='/assets/noimage/mini.png']"
+        @old_thumb_src = page.find("a.image-modal img")['src']
+
+        # Click image
+        page.find("a.image-modal").trigger('click')
+      end
+
+      # Shows upload modal
+      expect(page).to have_selector "div.reveal-modal.product-image-upload"
+
+      within "div.reveal-modal.product-image-upload" do
+        # Shows preview of current image
+        expect(page).to have_css "img.preview"
+        old_image_src = page.find("img.preview")['src']
+
+        # Upload a new image file
+        attach_file 'image-upload', Rails.root.join("public/500.jpg"), visible: false
+
+        # Shows spinner whilst loading
+        expect(page).to have_css "img.spinner", visible: true
+        expect(page).to_not have_css "img.spinner", visible: true
+
+        # Shows new image when finished
+        expect(page).to have_css "img.preview"
+        @new_image_src = page.find("img.preview")['src']
+        expect(old_image_src) != @new_image_src
+
+        # Close modal
+        page.find("a.close-reveal-modal").click
+      end
+
+      expect(page).to_not have_selector "div.reveal-modal.product-image-upload"
+
+      within "table#listing_products tr#p_#{product.id}" do
+        # New thumbnail is shown in image column
+        @new_thumb_src = page.find("a.image-modal img")['src']
+        expect(@old_thumb_src) != @new_thumb_src
+
+        page.find("a.image-modal").trigger('click')
+      end
+
+      expect(page).to have_selector "div.reveal-modal.product-image-upload"
+
+      within "div.reveal-modal.product-image-upload" do
+        # Upload another image file
+        attach_file 'image-upload', Rails.root.join("public/422.jpg"), visible: false
+
+        # Overwrites existing image
+        expect(page).to have_css "img.preview"
+        newer_image_src = page.find("img.preview")['src']
+        expect(@new_image_src) != newer_image_src
+
+        page.find("a.close-reveal-modal").click
+      end
+
+      within "table#listing_products tr#p_#{product.id}" do
+        # Newer thumbnail is shown in image column
+        newer_thumb_src = page.find("a.image-modal img")['src']
+        expect(@new_thumb_src) != newer_thumb_src
+      end
+    end
+  end
 end
```

Comment:
I'd be great to add a controller test also. IMO we rely way too much on integration tests and lack a proper testing pyramid.

So far I only see the happy path tested here, and that might make sense for an integration one.

Reply:
I've tested for both paths now in the spec. I used an integration spec instead of a controller spec because the file handling and response is all being done by `angularjs-file-upload`, so this controller is only receiving a pre-validated image from the file uploader, and is responding only to the file uploader.

What would you suggest for the controller spec?

Reply:
I strongly recommend you watch https://www.youtube.com/watch?v=kBOqaluDf2k and you'll have a better understanding of what I mean. We did that with @danypr92 yesterday and discussed examples from OFN and other projects we have. Very useful!